### PR TITLE
feat(custodian): add support for authorization for accessing keys

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Description

Adding custodian support. Adding support for Basic authentication on top of mTLS, to make sure that the keys being accessed belong to the respective party.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -128,6 +128,8 @@ pub struct Secrets {
     pub master_key: GcmAes256,
     #[cfg(feature = "aws")]
     pub kms_config: AwsKmsConfig,
+    pub access_token: masking::Secret<String>,
+    pub hash_context: masking::Secret<String>,
```
Comment (Chethan-rao) [inline src/config.rs:132]: needs to be added in `development.toml`?

File: src/core/crypto/custodian.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        .strip_prefix("Basic ")
+        .ok_or(ParsingError::ParsingFailed(
+            "Authorization scheme is not basic".to_string(),
+        ))?;
+    let credential = credential.trim();
+    let credential =
+        BASE64_ENGINE
+            .decode(credential)
+            .change_context(ParsingError::DecodingFailed(
+                "Failed while decoding base64".to_string(),
```
Comment (Chethan-rao) [inline src/core/crypto/custodian.rs:68]: Nit:

```suggestion
                "Failed while base64 decoding the authorization header".to_string(),
```

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: src/core/crypto/crux.rs [hunk 3]
```diff
@@ -99,11 +106,20 @@ impl DataEncrypter<EncryptedDataGroup> for DecryptedDataGroup {
...
         let key = GcmAes256::new(decrypted_key.key)?;
 
+        let stored_token = decrypted_key.token;
+        let provided_token = custodian.into_access_token(state);
+
+        ensure!(
+            stored_token.is_none() || (stored_token.eq(&provided_token)),
+            errors::CryptoError::AuthenticationFailed
+        );
+
```
Comment (dracarys18) [inline src/core/crypto/crux.rs:122]: Can you also make sure that the Locker Key identifier's are always authenticated

File: src/core/crypto/crux.rs [hunk 4]
```diff
@@ -125,12 +141,21 @@ impl DataDecrypter<DecryptedDataGroup> for EncryptedDataGroup {
...
         state: &AppState,
         identifier: &Identifier,
+        custodian: Custodian,
     ) -> errors::CustomResult<DecryptedDataGroup, errors::CryptoError> {
         let version = FxHashSet::from_iter(self.0.values().map(|d| d.version));
         let decrypted_keys = Key::get_multiple_keys(state, identifier, version)
             .await
             .switch()?;
 
+        let mut stored_tokens = decrypted_keys.values().map(|k| &k.token);
```
Comment (dracarys18) [inline src/core/crypto/crux.rs:151]:
```suggestion
        let stored_tokens = decrypted_keys.values().map(|k| &k.token);
```

File: src/core/crypto/crux.rs [hunk 5]
```diff
@@ -160,9 +185,19 @@ impl DataEncrypter<EncryptedData> for DecryptedData {
...
         let decrypted_key = Key::get_key(state, identifier, version).await.switch()?;
+
+        let stored_token = decrypted_key.token;
+        let provided_token = custodian.into_access_token(state);
+
+        ensure!(
+            stored_token.is_none() || (stored_token.eq(&provided_token)),
+            errors::CryptoError::AuthenticationFailed
+        );
+
```
Comment (dracarys18) [inline src/core/crypto/crux.rs:200]: Same here

File: src/core/crypto/crux.rs [hunk 6]
```diff
@@ -180,9 +215,19 @@ impl DataDecrypter<DecryptedData> for EncryptedData {
...
```
Comment (dracarys18) [inline src/core/crypto/crux.rs:230]: Same here

**Review by dracarys18 — CHANGES_REQUESTED** [review]

File: src/errors/crypto.rs [hunk 7]
```diff
@@ -19,6 +19,8 @@ pub enum CryptoError {
     ParseError(String),
     #[error("Invalid value")]
     InvalidValue,
+    #[error("Failed while authenticating the key")]
+    AuthenticationFailed,
```
Comment (Chethan-rao) [inline src/errors/crypto.rs:23]: Should this be mapped to `Unauthorized` in api error

File: src/core/crypto/crux.rs (hunk 4 above)
Comment (NishantJoshi00) [reply src/core/crypto/crux.rs:151]: It's required, as it's a opaque type that will be acted upon: when we call `.all` on it

**Comment by dracarys18** [comment]
Sure looks good to me, please test it once

**Review by dracarys18 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**